src/wp-includes/block-supports/block-style-variations.php

```diff
@@ -422,3 +407,93 @@ function wp_enqueue_block_style_variation_styles() {
+
+	foreach ( $variations as $key => $variation ) {
+		$supported_blocks = $variation['blockTypes'] ?? array();
+
+		/*
+		 * Standalone theme.json partial files for block style variations
+		 * will have their styles under a top-level property by the same name.
+		 * Variations defined within an existing theme.json or theme style
+		 * variation will themselves already be the required styles data.
+		 */
+		$variation_data = $variation['styles'] ?? $variation;
+
+		if ( empty( $variation_data ) ) {
+			continue;
+		}
+
+		/*
+		 * Block style variations read in via standalone theme.json partials
+		 * need to have their name set to the kebab case version of their title.
+		 */
+		$variation_name  = $have_named_variations ? $key : _wp_to_kebab_case( $variation['title'] );
+		$variation_label = $variation['title'] ?? $variation_name;
+
+		foreach ( $supported_blocks as $block_type ) {
+			$registered_styles = $registry->get_registered_styles_for_block( $block_type );
+
+			// Register block style variation if it hasn't already been registered.
+			if ( ! array_key_exists( $variation_name, $registered_styles ) ) {
+				register_block_style(
+					$block_type,
+					array(
+						'name'  => $variation_name,
+						'label' => $variation_label,
+					)
+				);
+			}
+		}
+	}
+}
+
+/**
+ * @access private
+ *
+ * Register the block style variations defined by the theme.
+ * These can come in three forms:
+ *
+ * - the theme's theme.json
+ * - the theme's partials (standalone files in styles that only define block style variations)
+ * - the user's theme.json (for example, theme style variations the user selected)
+ *
+ */
+function wp_register_block_style_variations_from_theme() {
+	// Partials.
+	$variations_partials = WP_Theme_JSON_Resolver::get_style_variations( 'block' );
+	wp_register_block_style_variations_from_theme_json_data( $variations_partials );
+
+	/*
+	 * Pull the data from the specific origin instead of the merged data.
+	 * This is because, for 6.6, we only support registering block style variations
+	 * for the 'theme' and 'custom' origins but not for 'default' (core theme.json)
+	 * or 'custom' (theme.json in a block).
+	 *
+	 * When/If we add support for every origin, we should switch to using the public API
+	 * instead, e.g.: wp_get_global_styles( array( 'blocks', 'variations' ) ).
+	 */
+
+	// theme.json of the theme.
+	$theme_json_theme = WP_Theme_JSON_Resolver::get_theme_data();
+	$variations_theme = $theme_json_theme->get_data()['styles']['blocks']['variations'] ?? array();
+	wp_register_block_style_variations_from_theme_json_data( $variations_theme );
+
+	// User data linked for this theme.
+	$theme_json_user = WP_Theme_JSON_Resolver::get_user_data();
+	$variations_user = $theme_json_user->get_data()['styles']['blocks']['variations'] ?? array();
+	wp_register_block_style_variations_from_theme_json_data( $variations_user );
+}
+add_filter( 'init', 'wp_register_block_style_variations_from_theme' );
```

Comment:
```suggestion
add_action( 'init', 'wp_register_block_style_variations_from_theme' );
```

Should this be `add_action`?

Reply:
Yes :) Fixed.